File: common/src/main/java/net/impactdev/gts/common/messaging/messages/listings/ClaimMessageImpl.java
package net.impactdev.gts.common.messaging.messages.listings;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.impactdev.gts.api.communication.message.errors.ErrorCode;
import net.impactdev.gts.api.communication.message.errors.ErrorCodes;
import net.impactdev.gts.api.communication.message.type.listings.ClaimMessage;
import net.impactdev.gts.api.util.TriState;
import net.impactdev.gts.common.messaging.GTSMessagingService;
import net.impactdev.gts.common.messaging.messages.AbstractMessage;
import net.impactdev.gts.common.plugin.GTSPlugin;
import net.impactdev.impactor.api.json.factory.JArray;
import net.impactdev.impactor.api.json.factory.JObject;
import net.impactdev.impactor.api.builders.Builder;
import net.impactdev.impactor.api.utilities.printing.PrettyPrinter;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public abstract class ClaimMessageImpl extends AbstractMessage implements ClaimMessage {

    protected final UUID listing;
    protected final UUID actor;
    protected final UUID receiver;
    protected final boolean auction;

    protected ClaimMessageImpl(UUID id, UUID listing, UUID actor, @Nullable UUID receiver, boolean auction) {
        super(id);
        this.listing = listing;
        this.actor = actor;
        this.receiver = receiver;
        this.auction = auction;
    }

    @Override
    public @NonNull UUID getListingID() {
        return this.listing;
    }

    @Override
    public @NonNull UUID getActor() {
        return this.actor;
    }

    @Override
    public Optional<UUID> getReceiver() {
        return Optional.ofNullable(this.receiver);
    }

    @Override
    public boolean isAuction() {
        return this.auction;
    }

    @Override
    public void print(PrettyPrinter printer) {
        printer.kv("Message ID", this.getID())
                .kv("Listing ID", this.listing)
                .kv("Actor", this.actor)
                .consume(p -> this.getReceiver().ifPresent(r -> p.kv("Receiver", r)))
                .kv("Is Auction", this.auction);
    }

    public static class ClaimRequestImpl extends ClaimMessageImpl implements ClaimMessage.Request {

        public static final String TYPE = "Listing/Claim/Request";

        public static ClaimRequestImpl decode(@Nullable JsonElement content, UUID id) {
            if (content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));
            UUID receiver = Optional.ofNullable(raw.get("receiver"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElse(null);
            boolean auction = Optional.ofNullable(raw.get("auction"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate auction check field"));
            return new ClaimRequestImpl(id, listing, actor, receiver, auction);
        }

        public ClaimRequestImpl(UUID id, UUID listing, UUID actor, @Nullable UUID receiver, boolean auction) {
            super(id, listing, actor, receiver, auction);
        }

        @Override
        public CompletableFuture<ClaimMessage.Response> respond() {
            // Divert to storage system to respond, there it should response with either the base
            // response type, or one specifically meant for auctions

            return GTSPlugin.instance().storage().processClaimRequest(this);
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .consume(o -> {
                                this.getReceiver().ifPresent(r -> o.add("receiver", r.toString()));
                            })
                            .add("auction", this.auction)
                            .toJson()
            );
        }
    }

    public static class ClaimResponseImpl extends ClaimMessageImpl implements ClaimMessage.Response {

        public static final String TYPE = "Listing/Claim/Response";

        public static ClaimMessage.Response decode(@Nullable JsonElement content, UUID id) {
            if (content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID request = Optional.ofNullable(raw.get("request"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate or parse request ID"));
            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));
            UUID receiver = Optional.ofNullable(raw.get("receiver"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElse(null);
            boolean auction = Optional.ofNullable(raw.get("auction"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate auction check field"));

            boolean successful = Optional.ofNullable(raw.get("successful"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate successful status"));
            ErrorCode error = Optional.ofNullable(raw.get("error"))
                    .map(x -> ErrorCodes.get(x.getAsInt()))
                    .orElse(null);

            ClaimResponseBuilder builder = ClaimResponseImpl.builder()
                    .id(id)
                    .request(request)
                    .listing(listing)
                    .actor(actor)
                    .receiver(receiver)
                    .error(error);
            if(successful) {
                builder.successful();
            }

            if(auction) {
                AuctionClaimResponseImpl.AuctionClaimResponseBuilder auc = AuctionClaimResponseImpl.builder().from(builder.build());

                boolean lister = Optional.ofNullable(raw.get("lister"))
                        .map(JsonElement::getAsBoolean)
                        .orElseThrow(() -> new IllegalStateException("Failed to locate lister status"));
                boolean winner = Optional.ofNullable(raw.get("winner"))
                        .map(JsonElement::getAsBoolean)
                        .orElseThrow(() -> new IllegalStateException("Failed to locate winner status"));
                Map<UUID, Boolean> others = Optional.ofNullable(raw.get("others"))
                        .map(element -> {
                            Map<UUID, Boolean> result = Maps.newHashMap();
                            JsonArray array = element.getAsJsonArray();
                            for(JsonElement entry : array) {
                                entry.getAsJsonObject().entrySet().forEach(e -> {
                                    result.put(UUID.fromString(e.getKey()), e.getValue().getAsBoolean());
                                });
                            }

                            return result;
                        })
                        .orElseThrow(() -> new IllegalStateException("Failed to locate others status"));

                return auc.lister(lister).winner(winner).others(others).build();
            } else {
                return builder.build();
            }
        }

        protected final UUID request;

        protected final boolean successful;
        protected final ErrorCode error;

        protected long responseTime;

        public ClaimResponseImpl(ClaimResponseBuilder builder) {
            super(builder.id, builder.listing, builder.actor, builder.actor, builder instanceof AuctionClaimResponseImpl.AuctionClaimResponseBuilder);
            this.request = builder.request;
            this.successful = builder.successful;
            this.error = builder.error;
        }

        @Override
        public UUID getRequestID() {
            return this.request;
        }

        @Override
        public long getResponseTime() {
            return this.responseTime;
        }

        @Override
        public void setResponseTime(long millis) {
            this.responseTime = millis;
        }

        @Override
        public boolean wasSuccessful() {
            return this.successful;
        }

        @Override
        public Optional<ErrorCode> getErrorCode() {
            return Optional.ofNullable(this.error);
        }

        @Override
        public void print(PrettyPrinter printer) {
            super.print(printer);

            printer.newline().kv("Request ID", this.request);
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("request", this.getRequestID().toString())
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .consume(o -> {
                                this.getReceiver().ifPresent(r -> o.add("receiver", r.toString()));
                            })
                            .add("auction", this.auction)
                            .add("successful", this.successful)
                            .consume(o -> this.getErrorCode().ifPresent(e -> o.add("error", e.ordinal())))
                            .toJson()
            );
        }

        public static final class AuctionClaimResponseImpl extends ClaimResponseImpl implements ClaimMessage.Response.AuctionResponse {

            private final boolean lister;
            private final boolean winner;

            private final Map<UUID, Boolean> others;

            public AuctionClaimResponseImpl(AuctionClaimResponseBuilder builder) {
                super(builder);
                this.lister = builder.lister;
                this.winner = builder.winner;
                this.others = builder.others;
            }

            @Override
            public boolean hasListerClaimed() {
                return this.lister;
            }

            @Override
            public boolean hasWinnerClaimed() {
                return this.winner;
            }

            @Override
            public TriState hasOtherBidderClaimed(UUID uuid) {
                return Optional.ofNullable(this.others.get(uuid))
                        .map(TriState::fromBoolean)
                        .orElse(TriState.UNDEFINED);
            }

            @Override
            public List<UUID> getAllOtherClaimers() {
                return Lists.newArrayList(this.others.keySet());
            }

            @Override
            public void print(PrettyPrinter printer) {
                super.print(printer);
                printer.newline();
                printer.add("Lister Claimed: " + this.lister);
                printer.add("Winner Claimed: " + this.winner);
                printer.add("Others:");
                for(UUID id : this.others.keySet()) {
                    printer.add("  - " + id.toString());

                }
            }

            @Override
            public @NonNull String asEncodedString() {
                return GTSMessagingService.encodeMessageAsString(
                        TYPE,
                        this.getID(),
                        new JObject()
                                .add("request", this.getRequestID().toString())
                                .add("listing", this.listing.toString())
                                .add("actor", this.actor.toString())
                                .consume(o -> {
                                    this.getReceiver().ifPresent(r -> o.add("receiver", r.toString()));
                                })
                                .add("auction", this.auction)
                                .add("lister", this.lister)
                                .add("winner", this.winner)
                                .consume(o -> {
                                    JArray others = new JArray();
                                    this.others.forEach((user, state) -> {
                                        others.add(new JObject().add(user.toString(), state));
                                    });
                                    o.add("others", others);
                                })
                                .add("successful", this.successful)
                                .consume(o -> this.getErrorCode().ifPresent(e -> o.add("error", e.ordinal())))
                                .toJson()
                );
            }

            public static AuctionClaimResponseBuilder builder() {
                return new AuctionClaimResponseBuilder();
            }

            public static final class AuctionClaimResponseBuilder extends ClaimResponseBuilder {

                private boolean lister;
                private boolean winner;

                private Map<UUID, Boolean> others = Maps.newHashMap();

                public AuctionClaimResponseBuilder lister(boolean state) {
                    this.lister = state;
                    return this;
                }

                public AuctionClaimResponseBuilder winner(boolean state) {
                    this.winner = state;
                    return this;
                }

                public AuctionClaimResponseBuilder others(Map<UUID, Boolean> others) {
                    this.others = others;
                    return this;
                }

                public AuctionClaimResponseBuilder from(ClaimMessage.Response response) {
                    this.id = response.getID();
                    this.listing = response.getListingID();
                    this.actor = response.getActor();
                    this.receiver = response.getReceiver().orElse(null);

                    this.request = response.getRequestID();
                    this.successful = response.wasSuccessful();
                    this.error = response.getErrorCode().orElse(null);

                    return this;
                }

                @Override
                public AuctionClaimResponseImpl build() {
                    return new AuctionClaimResponseImpl(this);
                }

            }
        }

        public static ClaimResponseBuilder builder() {
            return new ClaimResponseBuilder();
        }

        public static class ClaimResponseBuilder implements Builder<ClaimMessage.Response> {

            protected UUID id;
            protected UUID listing;
            protected UUID actor;
            protected UUID receiver;

            protected UUID request;
            protected boolean successful;
            protected ErrorCode error;

            public ClaimResponseBuilder id(UUID id) {
                this.id = id;
                return this;
            }

            public ClaimResponseBuilder listing(UUID listing) {
                this.listing = listing;
                return this;
            }

            public ClaimResponseBuilder actor(UUID actor) {
                this.actor = actor;
                return this;
            }

            public ClaimResponseBuilder receiver(UUID receiver) {
                this.receiver = receiver;
                return this;
            }

            public AuctionClaimResponseImpl.AuctionClaimResponseBuilder auction() {
                return AuctionClaimResponseImpl.builder().from(this.build());
            }

            public ClaimResponseBuilder request(UUID request) {
                this.request = request;
                return this;
            }

            public ClaimResponseBuilder successful() {
                this.successful = true;
                return this;
            }

            public ClaimResponseBuilder error(ErrorCode error) {
                this.error = error;
                return this;
            }

            @Override
            public ClaimResponseImpl build() {
                return new ClaimResponseImpl(this);
            }
        }
    }

}

File: common/src/main/kotlin/net/impactdev/gts/common/messaging/messages/listings/ClaimMessageImpl.kt
package net.impactdev.gts.common.messaging.messages.listings

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import com.google.gson.JsonElement
import net.impactdev.gts.api.messaging.message.errors.ErrorCode
import net.impactdev.gts.api.messaging.message.errors.ErrorCodes
import net.impactdev.gts.api.messaging.message.type.listings.ClaimMessage
import net.impactdev.gts.api.messaging.message.type.listings.ClaimMessage.Response.AuctionResponse
import net.impactdev.gts.api.util.PrettyPrinter
import net.impactdev.gts.api.util.TriState
import net.impactdev.gts.common.messaging.GTSMessagingService
import net.impactdev.gts.common.messaging.messages.AbstractMessage
import net.impactdev.gts.common.messaging.messages.listings.ClaimMessageImpl.ClaimResponseImpl.AuctionClaimResponseImpl.AuctionClaimResponseBuilder
import net.impactdev.gts.common.plugin.GTSPlugin
import net.impactdev.impactor.api.json.factory.JArray
import net.impactdev.impactor.api.json.factory.JObject
import net.impactdev.impactor.api.utilities.Builder
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Consumer
import kotlin.collections.List
import kotlin.collections.Map
import kotlin.collections.MutableMap
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.forEach
import kotlin.collections.get
import kotlin.collections.set

abstract class ClaimMessageImpl protected constructor(
    id: UUID?,
    override val listingID: UUID,
    override val actor: UUID,
    protected override val receiver: UUID?,
    override val isAuction: Boolean
) : AbstractMessage(
    id!!
), ClaimMessage {
    override fun getReceiver(): Optional<UUID> {
        return Optional.ofNullable(receiver)
    }

    override fun print(printer: PrettyPrinter) {
        printer.kv("Message ID", id)
            .kv("Listing ID", listingID)
            .kv("Actor", actor)
            .consume(Consumer { p: PrettyPrinter? -> getReceiver().ifPresent { r: UUID? -> p!!.kv("Receiver", r) } })
            .kv("Is Auction", isAuction)
    }

    class ClaimRequestImpl(id: UUID?, listing: UUID, actor: UUID, receiver: UUID?, auction: Boolean) :
        ClaimMessageImpl(id, listing, actor, receiver, auction), ClaimMessage.Request {
        override fun respond(): CompletableFuture<ClaimMessage.Response?>? {
            // Divert to storage system to respond, there it should response with either the base
            // response type, or one specifically meant for auctions
            return GTSPlugin.Companion.getInstance().getStorage().processClaimRequest(this)
        }

        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .consume { o: JObject -> getReceiver().ifPresent { r: UUID -> o.add("receiver", r.toString()) } }
                    .add("auction", isAuction)
                    .toJson()
            )
        }

        companion object {
            const val TYPE = "Listing/Claim/Request"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID?): ClaimRequestImpl {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                val receiver = Optional.ofNullable(raw["receiver"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElse(null)
                val auction = Optional.ofNullable(raw["auction"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate auction check field") }
                return ClaimRequestImpl(id, listing, actor, receiver, auction)
            }
        }
    }

    open class ClaimResponseImpl(builder: ClaimResponseBuilder) : ClaimMessageImpl(
        builder.id,
        builder.listing!!,
        builder.actor!!,
        builder.actor,
        builder is AuctionClaimResponseBuilder
    ), ClaimMessage.Response {
        override val requestID: UUID?
        protected val successful: Boolean
        protected val error: ErrorCode?
        override var responseTime: Long = 0
        override fun wasSuccessful(): Boolean {
            return successful
        }

        override val errorCode: Optional<ErrorCode>
            get() = Optional.ofNullable(error)

        override fun print(printer: PrettyPrinter) {
            super.print(printer)
            printer.add().kv("Request ID", requestID)
        }

        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("request", requestID.toString())
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .consume { o: JObject -> getReceiver().ifPresent { r: UUID -> o.add("receiver", r.toString()) } }
                    .add("auction", isAuction)
                    .add("successful", successful)
                    .consume { o: JObject -> errorCode.ifPresent { e: ErrorCode -> o.add("error", e.ordinal()) } }
                    .toJson()
            )
        }

        class AuctionClaimResponseImpl(builder: AuctionClaimResponseBuilder) : ClaimResponseImpl(builder),
            AuctionResponse {
            private val lister: Boolean
            private val winner: Boolean
            private val others: Map<UUID, Boolean>
            override fun hasListerClaimed(): Boolean {
                return lister
            }

            override fun hasWinnerClaimed(): Boolean {
                return winner
            }

            override fun hasOtherBidderClaimed(uuid: UUID?): TriState? {
                return Optional.ofNullable(others[uuid])
                    .map { obj: Boolean? -> TriState.fromBoolean() }
                    .orElse(TriState.UNDEFINED)
            }

            override val allOtherClaimers: List<UUID>
                get() = Lists.newArrayList(others.keys)

            override fun print(printer: PrettyPrinter) {
                super.print(printer)
                printer.add()
                printer.add("Lister Claimed: " + lister)
                printer.add("Winner Claimed: " + winner)
                printer.add("Others:")
                for (id in others.keys) {
                    printer.add("  - $id")
                }
            }

            override fun asEncodedString(): String {
                return GTSMessagingService.Companion.encodeMessageAsString(
                    TYPE,
                    id,
                    JObject()
                        .add("request", requestID.toString())
                        .add("listing", listingID.toString())
                        .add("actor", actor.toString())
                        .consume { o: JObject ->
                            getReceiver().ifPresent { r: UUID ->
                                o.add(
                                    "receiver",
                                    r.toString()
                                )
                            }
                        }
                        .add("auction", isAuction)
                        .add("lister", lister)
                        .add("winner", winner)
                        .consume { o: JObject ->
                            val others = JArray()
                            this.others.forEach { (user: UUID, state: Boolean?) ->
                                others.add(
                                    JObject().add(
                                        user.toString(),
                                        state
                                    )
                                )
                            }
                            o.add("others", others)
                        }
                        .add("successful", successful)
                        .consume { o: JObject -> errorCode.ifPresent { e: ErrorCode -> o.add("error", e.ordinal()) } }
                        .toJson()
                )
            }

            class AuctionClaimResponseBuilder : ClaimResponseBuilder() {
                var lister = false
                var winner = false
                var others: Map<UUID, Boolean> = Maps.newHashMap()
                fun lister(state: Boolean): AuctionClaimResponseBuilder {
                    lister = state
                    return this
                }

                fun winner(state: Boolean): AuctionClaimResponseBuilder {
                    winner = state
                    return this
                }

                fun others(others: Map<UUID, Boolean>): AuctionClaimResponseBuilder {
                    this.others = others
                    return this
                }

                override fun from(response: ClaimMessage.Response): AuctionClaimResponseBuilder? {
                    id = response.iD
                    listing = response.listingID
                    actor = response.actor
                    receiver = response.receiver!!.orElse(null)
                    request = response.requestID
                    successful = response.wasSuccessful()
                    error = response.errorCode.orElse(null)
                    return this
                }

                override fun build(): AuctionClaimResponseImpl {
                    return AuctionClaimResponseImpl(this)
                }
            }

            companion object {
                fun builder(): AuctionClaimResponseBuilder {
                    return AuctionClaimResponseBuilder()
                }
            }

            init {
                lister = builder.lister
                winner = builder.winner
                others = builder.others
            }
        }

        open class ClaimResponseBuilder : Builder<ClaimMessage.Response, ClaimResponseBuilder?> {
            var id: UUID? = null
            var listing: UUID? = null
            var actor: UUID? = null
            protected var receiver: UUID? = null
            var request: UUID? = null
            var successful = false
            var error: ErrorCode? = null
            fun id(id: UUID?): ClaimResponseBuilder {
                this.id = id
                return this
            }

            fun listing(listing: UUID?): ClaimResponseBuilder {
                this.listing = listing
                return this
            }

            fun actor(actor: UUID?): ClaimResponseBuilder {
                this.actor = actor
                return this
            }

            fun receiver(receiver: UUID?): ClaimResponseBuilder {
                this.receiver = receiver
                return this
            }

            fun auction(): AuctionClaimResponseBuilder? {
                return AuctionClaimResponseImpl.builder().from(build())
            }

            fun request(request: UUID?): ClaimResponseBuilder {
                this.request = request
                return this
            }

            fun successful(): ClaimResponseBuilder {
                successful = true
                return this
            }

            fun error(error: ErrorCode?): ClaimResponseBuilder {
                this.error = error
                return this
            }

            override fun from(response: ClaimMessage.Response): ClaimResponseBuilder? {
                return null
            }

            override fun build(): ClaimResponseImpl {
                return ClaimResponseImpl(this)
            }
        }

        companion object {
            const val TYPE = "Listing/Claim/Response"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID?): ClaimMessage.Response {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val request = Optional.ofNullable(raw["request"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate or parse request ID") }
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                val receiver = Optional.ofNullable(raw["receiver"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElse(null)
                val auction = Optional.ofNullable(raw["auction"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate auction check field") }
                val successful = Optional.ofNullable(raw["successful"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate successful status") }
                val error = Optional.ofNullable(raw["error"])
                    .map { x: JsonElement -> ErrorCodes[x.asInt] }
                    .orElse(null)
                val builder = builder()
                    .id(id)
                    .request(request)
                    .listing(listing)
                    .actor(actor)
                    .receiver(receiver)
                    .error(error)
                if (successful) {
                    builder.successful()
                }
                return if (auction) {
                    val auc = AuctionClaimResponseImpl.builder().from(builder.build())
                    val lister = Optional.ofNullable(raw["lister"])
                        .map { obj: JsonElement -> obj.asBoolean }
                        .orElseThrow { IllegalStateException("Failed to locate lister status") }
                    val winner = Optional.ofNullable(raw["winner"])
                        .map { obj: JsonElement -> obj.asBoolean }
                        .orElseThrow { IllegalStateException("Failed to locate winner status") }
                    val others = Optional.ofNullable(raw["others"])
                        .map<Map<UUID, Boolean>> { element: JsonElement ->
                            val result: MutableMap<UUID, Boolean> = Maps.newHashMap()
                            val array = element.asJsonArray
                            for (entry in array) {
                                entry.asJsonObject.entrySet().forEach(Consumer { e: Map.Entry<String?, JsonElement> ->
                                    result[UUID.fromString(e.key)] = e.value.asBoolean
                                })
                            }
                            result
                        }
                        .orElseThrow { IllegalStateException("Failed to locate others status") }
                    auc!!.lister(lister).winner(winner).others(others).build()
                } else {
                    builder.build()
                }
            }

            fun builder(): ClaimResponseBuilder {
                return ClaimResponseBuilder()
            }
        }

        init {
            requestID = builder.request
            successful = builder.successful
            error = builder.error
        }
    }
}